Fix pack size related marshal issues.

Akarinnnnn opened the pull request:
Our goal is to fix most callback system related marshal issues.

Added a conditional marshalling table to decide which alignment to use at runtime, this should fix most callback-retrieve related issues.

Struct parameters in interface methods are ~~still~~ need to fix. Edit: Completed.

I think we ~~still~~ need some effort to fix parameters, maybe we still have to ship different assembly for different platforms.
Edit: Parameter marshal is done, through it looks ugly.

Now the only problem is to find out which struct we treat it specially, in generation process.

rlabrecque committed:
Import steamworksparser.py

rlabrecque committed:
Added support for unnamed enums (enum constants)

rlabrecque committed:
Added .gitignore for *.pyc files

rlabrecque committed:
The GameServer APIContext now has a UGC accessor

rlabrecque committed:
Fixed typedef.filename always being blank

rlabrecque committed:
Enable recursive parsing for multiple /**/ comment blocks on one line

rlabrecque committed:
Skip _STEAM_CALLBACK_ Macros

rlabrecque committed:
Added README

rlabrecque committed:
Added MIT license

rlabrecque committed:
ArgAttribute improvements

rlabrecque committed:
Added parsing FunctionAttributes

rlabrecque committed:
Parse STEAM_PRIVATE_API()

rlabrecque committed:
Skip parsing #pragma and #undef

And sorted Skippedlines

rlabrecque committed:
Allow setting ArgAttribute upon creation

rlabrecque committed:
Fixed loading CP-1252 files on linux...

rlabrecque committed:
Added typedefs to comments and made typedefs exist per file.

rlabrecque committed:
Updated GameServerInterfaces

rlabrecque committed:
Updated for Steamworks 1.42

rlabrecque committed:
Updated for Steamworks 1.43

rlabrecque committed:
Update for 1.44, skipping SteamNetworkingSockets for now

rlabrecque committed:
Fix attribute value parsing

rlabrecque committed:
Ignore steamnetworkingsockets.h

rlabrecque committed:
Update for Steamworks SDK 1.48

rlabrecque committed:
Add some support for steamnetworkingtypes.h

rlabrecque committed:
Add support for isteamnetworkingutils.h

rlabrecque committed:
Add support for steamnetworkingsockets.h and steamdatagram_tickets.h

rlabrecque committed:
Add game server versions of isteamnetworkingutils and isteamnetworkingsockets

rlabrecque committed:
Stop parsing isteamcontroller.h

rlabrecque committed:
Update README and LICENSE

rlabrecque committed:
Correctly parse inline function definitions inside interfaces

rlabrecque committed:
Update for Steamworks SDK 1.52

rlabrecque committed:
Update for Steamworks SDK 1.52 - sort-of handle non-callback enum constants inside a struct

rlabrecque committed:
Update for Steamworks SDK 1.55 - skip isteamdualsense.h

TwoPoint-AndyChappell committed:
Add support for the new STEAM_CALLBACK_MEMBER_ARRAY macro

TwoPoint-AndyChappell committed:
Handle the static const int value within GetTicketForWebApiResponse_t
correctly as a constant.  Scan things at scope depth 1 as well as 0 now
to catch it, and also exclude field names in structs that contain an =.
Means the field is no longer added to the struct and is instead added
correctly to constants.

rlabrecque committed:
Add SteamNetworkingMessages as a GameServer Interface

Chicken-Bones commented:
I was thinking something more like
```cs
namespace Callbacks;

[StructLayout(LayoutKind.Sequential, Pack = 4)]
[CallbackIdentity(Constants.k_iSteamUGCCallbacks + 6)]
public struct DownloadItemResult_t {
	public const int k_iCallback = Constants.k_iSteamUGCCallbacks + 6;
	public AppId_t m_unAppID;
	public PublishedFileId_t m_nPublishedFileId;
	public EResult m_eResult;
}


internal class NativeLargePack
{
	[StructLayout(LayoutKind.Sequential, Pack = 8)]
	internal struct DownloadItemResult_t {
		public AppId_t m_unAppID;
		public PublishedFileId_t m_nPublishedFileId;
		public EResult m_eResult;
		
		public static implicit operator DownloadItemResult_t(Callbacks.DownloadItemResult_t o) => new() {
			m_unAppID = o.m_unAppID,
			m_nPublishedFileId = o.m_nPublishedFileId,
			m_eResult = o.m_eResult
		};
		
		public static implicit operator Callbacks.DownloadItemResult_t(DownloadItemResult_t o) => new() {
			m_unAppID = o.m_unAppID,
			m_nPublishedFileId = o.m_nPublishedFileId,
			m_eResult = o.m_eResult
		};
	}
	
	[DllImport("steam")]
	private static DownloadItemResult_t ImaginaryNativeFunction(DownloadItemResult_t arg);
}


internal class Native
{
	[DllImport("steam")]
	private static DownloadItemResult_t ImaginaryNativeFunction(DownloadItemResult_t arg);
}


public static DownloadItemResult_t ImaginaryNativeFunction(DownloadItemResult_t arg) {
	if (Packsize.IsLargePack)
		return NativeLargePack.ImaginaryNativeFunction(arg);
	
	return Native.ImaginaryNativeFunction(arg);
}
```

Where we only generate structs and functions in `NativeLargePack` when packsize actually matters

Akarinnnnn commented:
Bad news: most structs are packsize matters, only few structs specified packsize.

Chicken-Bones commented:
I ran `Marsha.SizeOf` on all structs in steamworks dll under `Packsize.value = 8` and `Packsize.value = 4`
I find packsize matters for 87/302 structs. 
81/87 are callback structs with no direct references

Only ones used in `NativeMethods.cs` are
- `SteamPartyBeaconLocation_t`
- `SteamUGCDetails_t`
- `LeaderboardEntry_t`

Special cases are:
- `CallbackMsg_t` Manually marshalled in `CallbackDispatcher`
- `SteamParamStringArray_t` Manually marshalled in `InteropHelp`
- `ValvePackingSentinel_t`

So... only need to generate packsize dependent `NativeMethods` wrappers for small number of functions. I think it's possible to autodetect when packsize matters in python and only generate for such types. I suggest implicit conversions because they make it easy to autogenerate native wrapper methods.

Btw, for `ConditionalMarshallerTable.Marshal` you can abuse static generic specialization to get more typesafe impl. With proposed implicit conversions in `_LargePack` struct variants:

```cs
	internal static partial class ConditionalMarshallerTable
	{
		private static class Impl<T>
		{
			public static Func<IntPtr, T> Marshal = System.Runtime.InteropServices.Marshal.PtrToStructure<T>;
		}

		public static T Marshal<T>(IntPtr unmanagetype) => Impl<T>.Marshal(unmanagetype);

		// partial, in generated file
		static ConditionalMarshallerTable()
		{
			if (!Packsize.IsLargePack)
				return;

			Impl<DlcInstalled_t>.Marshal = (unmanaged) => System.Runtime.InteropServices.Marshal.PtrToStructure<DlcInstalled_t_LargePack>(unmanaged);
			Impl<NewUrlLaunchParameters_t>.Marshal = (unmanaged) => System.Runtime.InteropServices.Marshal.PtrToStructure<NewUrlLaunchParameters_t_LargePack>(unmanaged);
		}
	}
```

Akarinnnnn commented:
Do you mean we add memory layout analysis code in codegen? Finding these special structs out needs knowledge of which location does fields stored, I think.

Static generic specialization is a good idea. 

`ValvePackingSentinel_t` has no meaningful usage, we can ignore it. It just used to test is pack-size correct at runtime. In Any CPU I removed this test.

Akarinnnnn commented:
Now I will provide a temporary fix by packing different assemblies into one NuGet package. Hope we can fix quicker.

Chicken-Bones commented:
> Do you mean we add memory layout analysis code in codegen? Finding these special structs out needs knowledge of which location does fields stored, I think.

Yes, the rules of determining field offset is relatively simple for recursive algorithm. 

This is my (untested) attempt at some pseudocode:
```py
def get_size_and_alignment(type, pack):
    if is_primitive(type):
        size = g_PrimitiveSizeTable[type] # similar to g_TypeConversionDict, note that all pointer types will be size and alignment 8
        return size, size

    # maybe load pack = g_CustomPackSize here
    offset = 0
    align = 1
    for field in type.fields:
         (f_size, f_align) = get_size_and_alignment(field, pack)
         f_align = min(f_align, pack)
         align = max(align, f_align)
         offset += (f_align - offset) % f_align # align this field
         offset += f_size
    
    offset += (align - offset) % align # pad struct size to a multiple of largest field alignment
    return offset, align
```

Akarinnnnn committed:
Handle UTF-8 encoded headers better

Akarinnnnn commented:
@Chicken-Bones I got an algorithm but can't sure whether it correct. Would you like to share an effected struct name and field offsets for test case? Provide both large and small pack is better.

Offset calculation API in my design is `struct.struct_layout(packOverride)`, returns a list like `[("eResult", 4, 4), ("bSomeFlag", 5, 1)]`

Chicken-Bones commented:
These files contain the total sizes of the structs for different packsize.

[4.txt](https://github.com/user-attachments/files/22937367/4.txt)
[8.txt](https://github.com/user-attachments/files/22937368/8.txt)

I didn't write the reflection or codegen for pulling all the field offsets, but that's probably unnecessary anyway, since any change in field offset involves adding padding which increases total struct size.

Akarinnnnn commented:
There is a good mews: I have been hired. But I may too busy to contribute this project since that. By the way, the computer stored this repository may not be bring to that city, which will cause I can only contribute this lib at holidays.


Anyway, good luck to us, and this fork.



---Original---
From: ***@***.***&gt;
Date: Thu, Oct 16, 2025 06:45 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [Akarinnnnn/Steamworks.NET.AnyCPU] Fix most callback systemrelated marshal issues. (PR #1)


Chicken-Bones left a comment (Akarinnnnn/Steamworks.NET.AnyCPU#1)
 
These files contain the total sizes of the structs for different packsize.
 
4.txt
 8.txt
 
I didn't write the reflection or codegen for pulling all the field offsets, but that's probably unnecessary anyway, since any change in field offset involves adding padding which increases total struct size.
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***&gt;

Chicken-Bones commented:
Congrats, and thanks for your work so far

Akarinnnnn committed:
Try add struct layout analysis, add type annotations

Akarinnnnn commented:
I met problem on nested struct `SteamInputActionEvent_t::AnalogAction_t` in `isteaminput.h`, currently they are in my parser fork. I'm considering to add a PR for upcoming commits. How do you think?

Chicken-Bones commented:
Yes, please update the PR with work in progress. Then show me how I can run/test and I will investigate error.

Akarinnnnn committed:
Attempt to make parser smarter

Adding support for nested types and unions, these work are in progress.

Akarinnnnn committed:
Attempt to calculate total size of structs, part 3

Currently I still have to debug unions, they caused nested type stack corrupt I think... I need help

Akarinnnnn committed:
Attempt to calculate total size of structs, part 4

Nested type parsing and some special structs are ok I think. Next i will work on calculation.

Akarinnnnn committed:
Attempt to calculate total size of structs, attempt 5

Is it work?

Akarinnnnn committed:
[Feature] Implement struct field offset analysis.

This feature can be used to generate helper structs that need to generate large and small pack variant. This problem is caused by default pack-size make the struct have different layout on different platforms.

Generate these struct with both variant is very important to building a unified Any CPU Steamworks.NET assembly. After this commit

Akarinnnnn committed:
Preparing use subtree instead of submodule

Akarinnnnn committed:
Add 'CodeGen/SteamworksParser/' from commit '665fb115a9182883457a6bb565d1c3b7c07a8c42'

git-subtree-dir: CodeGen/SteamworksParser
git-subtree-mainline: cffcf0484ed3d73872d6bd577be081dd36eed5e1
git-subtree-split: 665fb115a9182883457a6bb565d1c3b7c07a8c42

Akarinnnnn committed:
Make struct layout analysis work

Akarinnnnn committed:
Add support to hack marshalling on demand

This commit contains only codegen

Akarinnnnn committed:
Hide const packsize, expose it by static readonly field instead

Akarinnnnn committed:
Allow generator to ignore structs, clearify ambiguous type of `Field.arraysize` by Python type annotation

Akarinnnnn committed:
Source geneator now generate both native function variant for packsize aware structs.

Corresponding wrapper method is pending.

Akarinnnnn committed:
Source generator now generate both variant of packsize aware struct.

Akarinnnnn commented:
Currently we have callback structs like this. Main structs(`DownloadItemResult_t`) is exposed, large pack variant `DownloadItemResult_t_LargePack` is internal. They can do conversion by implicit conversion operators as you supposed. Your generic hacks also helped me a lot, thanks.
```csharp
	//-----------------------------------------------------------------------------
	// Purpose: result of DownloadItem(), existing item files can be accessed again
	//-----------------------------------------------------------------------------
	[StructLayout(LayoutKind.Sequential, Pack = Packsize.value)]
	[CallbackIdentity(Constants.k_iSteamUGCCallbacks + 6)]
	public struct DownloadItemResult_t
	#if STEAMWORKS_ANYCPU
		: ICallbackIdentity
	#endif
	{
		public const int k_iCallback = Constants.k_iSteamUGCCallbacks + 6;
		public static int CallbackIdentity { get; } = Constants.k_iSteamUGCCallbacks + 6;
		public AppId_t m_unAppID;
		public PublishedFileId_t m_nPublishedFileId;
		public EResult m_eResult;
	}

	#if STEAMWORKS_ANYCPU
	//-----------------------------------------------------------------------------
	// Purpose: result of DownloadItem(), existing item files can be accessed again
	//-----------------------------------------------------------------------------
	[StructLayout(LayoutKind.Sequential, Pack = 8)]
	[CallbackIdentity(Constants.k_iSteamUGCCallbacks + 6)]
	internal struct DownloadItemResult_t_LargePack {
		public AppId_t m_unAppID;
		public PublishedFileId_t m_nPublishedFileId;
		public EResult m_eResult;

		public static implicit operator DownloadItemResult_t(DownloadItemResult_t_LargePack value) {
			DownloadItemResult_t result = default;
			result.m_unAppID = value.m_unAppID;
			result.m_nPublishedFileId = value.m_nPublishedFileId;
			result.m_eResult = value.m_eResult;
			return result;
		}

		public static implicit operator DownloadItemResult_t_LargePack(DownloadItemResult_t value) {
			DownloadItemResult_t_LargePack result = default;
			result.m_unAppID = value.m_unAppID;
			result.m_nPublishedFileId = value.m_nPublishedFileId;
			result.m_eResult = value.m_eResult;
			return result;
		}
	}

	#endif
```

I also added an interface contains static callback id property. I want to use it to replace reflection based identity number acquisition.

Akarinnnnn committed:
Generated code not work, need fix.

There is a conflict constant `k_cchMaxString` caused by  `SteamNetworkingIdentity` and `SteamNetworkingIPAddr` defines same named field, in two different anonymous enum

Akarinnnnn committed:
Fixed constant value confliction by ignore a problemtic file.

Types inside struct and `SteamParamStringArray_t` may get fixed today(in UTC+8)

Akarinnnnn committed:
Fix steam headers parser for ignoring `SteamParamStringArray_t::m_ppStrings`

Akarinnnnn committed:
Fix `Callback<T>` doesn't marshal Any CPU correctly.

Interface parameter usage of struct still need work.

Akarinnnnn commented:
> Yes, please update the PR with work in progress. Then show me how I can run/test and I will investigate error.

SteamworksParser PR is here: https://github.com/Akarinnnnn/SteamworksParser/pull/2 . It seems to work but I can't determine is it really work.

I use VSCode for development tool. To debug repository *SteamworksParser* you can just hit F5, as python plugins are installed. Run or debug `debug_entrypoint.py` is the way how I debug, Also, you need to put steam headers into folder `SteamworksParser/steamtest/` to feed data for parser.

Chicken-Bones commented:
Unfortunately I did not find time to test/review in detail, and am going on holidays for a few weeks. 

One thing I did notice is that we still need to wrap the usages of 
```
SteamPartyBeaconLocation_t
SteamUGCDetails_t
LeaderboardEntry_t
```

in `NativeMethods` functions.

Akarinnnnn committed:
Make SrcGen can generate Any CPU interface wrapper.

Native methods are still in progress. First is LP/SP native function pair `ISteamParties_GetBeaconLocationData` has been duplicated twice, it seems same function generated twice.

Secondly large pack `ISteamParties_GetAvailableBeaconLocations`'s second arg, `p_locationList_lp`, 's associated type is not properly generated. Array brackets is generated before `_LargePack`. I' ll try to find out why.

Akarinnnnn commented:
Leaderboard and ugcdetail has been generated well. But Beacon[] is generated wrongly as Beacon[]_LargePack as commit message said. I'll try to gind out why.



---Original---
From: ***@***.***&gt;
Date: Thu, Oct 23, 2025 19:15 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [Akarinnnnn/Steamworks.NET.AnyCPU] Fix most callback systemrelated marshal issues. (PR #1)


Chicken-Bones left a comment (Akarinnnnn/Steamworks.NET.AnyCPU#1)
 
Unfortunately I did not find time to test/review in detail, and am going on holidays for a few weeks.
 
One thing I did notice is that we still need to wrap the usages of
 SteamPartyBeaconLocation_t SteamUGCDetails_t LeaderboardEntry_t  
in NativeMethods functions.
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***&gt;

Akarinnnnn committed:
Fixed pack-size aware typed parameter array generation.

Akarinnnnn committed:
Regenerate native imports with array fix.

Akarinnnnn committed:
Merge remote-tracking branch 'origin/master' into anycpu-packable-fix-align-by-conditionalmarshal

Akarinnnnn committed:
Fix gameserver interfaces generation

Akarinnnnn committed:
All Steam interface wrappers seems work.

Akarinnnnn committed:
Regenerate steam interface wrappers.

Akarinnnnn commented:
I have published new package version `2025.162.6-a` which build from this PR. You can try it out if you are interested.

Akarinnnnn commented:
It seems `git subtree` broke GitHub's diff preview page, I recommend clone this to compare changes locally.

Akarinnnnn commented:
> Unfortunately I did not find time to test/review in detail, and am going on holidays for a few weeks.
> 
> One thing I did notice is that we still need to wrap the usages of
> 
> ```
> SteamPartyBeaconLocation_t
> SteamUGCDetails_t
> LeaderboardEntry_t
> ```
> 
> in `NativeMethods` functions.

They seems wrapped properly in current code. I have tested `SteamUGCDetails_t` and seems work.

Akarinnnnn committed:
Fix array parameter usage of packsize aware structs.

Regenerated files are also included.

Akarinnnnn committed:
Merge remote-tracking branch 'upstream/master' into anycpu-packable-fix-align-by-conditionalmarshal

Akarinnnnn commented:
I've found that it's hard to review changes of `SteamworksParser` in https://github.com/Akarinnnnn/Steamworks.NET.AnyCPU/pull/1 because parser file is *displayed as a new file*. It makes **massive noise** to our review process. Now I will sync changes in subtree periodically to https://github.com/Akarinnnnn/SteamworksParser/pull/2 for **better review** experience.

ryan-linehan committed:
include STEAMWORKS_WIN constant

ryan-linehan committed:
add a fix to dynamically load assembly on windows

Akarinnnnn committed:
Revert adding of macro `STEAMWORKS_WIN`

To avoid breaking internal symbol usages, `STEAMWORKS_WIN` is dropped versus `STEAMWORKS_LIN_OSX`.

Akarinnnnn committed:
Changed the time to hook dll loading.

Now we can do our hook in any AssemblyLoacContext s. DLL loading issues in godot(editor?)  caused by multiple A.L.C.s should be fixed.

Akarinnnnn committed:
Move loader hook fix to source generator's template.

Akarinnnnn committed:
Move loader hook fix to source generator's template.

Akarinnnnn committed:
Change the place of DLL resolve hook (#6)

Fixed #5 by move inject DLL loader hook to `NativeLibrary` provided interception point. This allows we hook loading process in any ALC, which is especially important for Godot.

Akarinnnnn committed:
Fix anycpu large pack native function was leak.

Akarinnnnn committed:
Fix non-anycpu variant uses anycpu wrapper

It will cause complication failure.

Akarinnnnn committed:
Regenreate to fix AnyCPU wrappers leak.

ryan-linehan committed:
update sendsignal to be pack aware with manual edits

ryan-linehan committed:
update structs.py to include SteamNetConnectionInfo_t as a sequential struct

ryan-linehan committed:
update send signal template to be pack size aware

ryan-linehan committed:
update nativemethods.txt to have pack aware dll import

Akarinnnnn committed:
Fix packsize aware struct nested in pack aware struct.

But pack aware struct nested in common struct are not in consideration. Write it down here for one day we met some bug related.

Chicken-Bones commented on com.rlabrecque.steamworks.net/Runtime/autogen/isteamgameserverugc.cs:
```diff
@@ -63,7 +63,21 @@ public static SteamAPICall_t SendQueryUGCRequest(UGCQueryHandle_t handle) {
 		/// </summary>
 		public static bool GetQueryUGCResult(UGCQueryHandle_t handle, uint index, out SteamUGCDetails_t pDetails) {
 			InteropHelp.TestIfAvailableGameServer();
+			#if STEAMWORKS_ANYCPU
```
better to invert this codegen perhaps?
```cs
			#if !STEAMWORKS_ANYCPU
			return NativeMethods.ISteamUGC_GetQueryUGCResult(CSteamGameServerAPIContext.GetSteamUGC(), handle, index, out pDetails);
			#else
			bool ret;
			if (!Packsize.IsLargePack) {
				ret = NativeMethods.ISteamUGC_GetQueryUGCResult(CSteamGameServerAPIContext.GetSteamUGC(), handle, index, out pDetails);
			} else {
				SteamUGCDetails_t pDetails_lp;
				ret = NativeMethods.ISteamUGC_GetQueryUGCResult(CSteamGameServerAPIContext.GetSteamUGC(), handle, index, out pDetails_lp);
				pDetails = pDetails_lp;
			}
			return ret;
			#endif
```

Akarinnnnn replied:
I initially choose to make `IsLargePack ==  true` first is because I wrote analyzer code and want use it earlier. Now for CoreCLR/CLR, unused branch will be completely eliminated, only scarification is bootstrap speed. But I'm not sure what will happen on Mono.

On the other hand, I think most client of this library will run on Windows, large pack one. By following put most happens case first in `if`s, I think this can reduce branch prediction miss.

Would you like to share the details about inverting codegen? I will here to listen.

Chicken-Bones replied:
I only meant to reduce the number of `#if` conditions by grouping all the any cpu code together including the `return ret`. It doesn't matter whether you use `#if STEAMWORKS_ANYCPU` or `#if !STEAMWORKS_ANYCPU`

Akarinnnnn replied:
Good idea!

Akarinnnnn replied:
I know why it looks so ugly, it's because `void Method(out string outStringParameter)` needs to assign the `out string`, but in native methods they are `Utf8StringHandle`. It's need to be instantized before *native* method invocation. By the way, this type needs explicit method invocation to extract a string. And this is why we can't just return after native method called, because `out string`s at this state is still *not initialized*.

If we generate `out string` assign pasta in `#else` block you suggested, we repeat assign generating ourself, which violates DRY. I think keep anycpu code ugly isn't so bad now.

Akarinnnnn replied:
This is an example for another type of operation before return:

https://github.com/Akarinnnnn/Steamworks.NET.AnyCPU/blob/e3fdc5e655cd62305fd07232bce0147e34d507b8/com.rlabrecque.steamworks.net/Runtime/autogen/isteammatchmaking.cs#L912-L930

Akarinnnnn replied:
I decide to publish this branch first, since it's function now. We can fix code style in next PR.

Chicken-Bones reviewed (dismissed):
I ran a `SizeOf` analysis on all structs and found many missing `_LargePack` types, and some unnecessarily generated `_LargePack` types.

[expected.txt](https://github.com/user-attachments/files/23579312/expected.txt)
[anycpu.txt](https://github.com/user-attachments/files/23579313/anycpu.txt)

Akarinnnnn commented:
This is a problem... Also there is struct in sockets context has unexpected pack size, see latest pr.



---Original---
From: ***@***.***&gt;
Date: Mon, Nov 17, 2025 17:59 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [Akarinnnnn/Steamworks.NET.AnyCPU] Fix pack size related marshalissues. (PR #1)




 
@Chicken-Bones requested changes on this pull request.
 
I ran a SizeOf analysis on all structs and found many missing _LargePack types, and some unnecessarily generated _LargePack types.
 
expected.txt
 anycpu.txt

 
In com.rlabrecque.steamworks.net/Runtime/autogen/isteamgameserverugc.cs:
 &gt; @@ -63,7 +63,21 @@ public static SteamAPICall_t SendQueryUGCRequest(UGCQueryHandle_t handle) {  		/// </summary&gt;  		public static bool GetQueryUGCResult(UGCQueryHandle_t handle, uint index, out SteamUGCDetails_t pDetails) {  			InteropHelp.TestIfAvailableGameServer(); +			#if STEAMWORKS_ANYCPU  
better to invert this codegen perhaps?
 			#if !STEAMWORKS_ANYCPU 			return NativeMethods.ISteamUGC_GetQueryUGCResult(CSteamGameServerAPIContext.GetSteamUGC(), handle, index, out pDetails); 			#else 			bool ret; 			if (!Packsize.IsLargePack) { 				ret = NativeMethods.ISteamUGC_GetQueryUGCResult(CSteamGameServerAPIContext.GetSteamUGC(), handle, index, out pDetails); 			} else { 				SteamUGCDetails_t pDetails_lp; 				ret = NativeMethods.ISteamUGC_GetQueryUGCResult(CSteamGameServerAPIContext.GetSteamUGC(), handle, index, out pDetails_lp); 				pDetails = pDetails_lp; 			} 			return ret; 			#endif
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were assigned.Message ID: ***@***.***&gt;

Akarinnnnn commented:
> [expected.txt](https://github.com/user-attachments/files/23579312/expected.txt) [anycpu.txt](https://github.com/user-attachments/files/23579313/anycpu.txt)

Hi would you upload these files again? I got `Not Found` error when I downloading it. Sharing it on GitHub Gists is also a good option.

Chicken-Bones commented:
https://gist.github.com/Chicken-Bones/bae093eca50ae2db3324174967bc7131

Akarinnnnn commented:
> I ran a `SizeOf` analysis on all structs and found many missing `_LargePack` types, and some unnecessarily generated `_LargePack` types.

I have to admit, our parser has a lot of limitations but Steamworks headers is not simple as I think, they have dependencies internally, which may confuse this parser and leads incorrect result.

Chicken-Bones commented:
Unfortunately this has turned out more complex than I envisioned. I'm willing to take a look at the parser myself but finding time will take longer. My apologies for the delay.

Akarinnnnn commented:
I have an idea about using C++ toolchain to find which one is really need to focus on. Since I have MSVC and WSL clang, why not?

Akarinnnnn commented:
https://gist.github.com/Akarinnnnn/0a35bb1d250505aa98020138498cfbe4

Here is my result(in progress). Line format is `"{name}\t{size}\t{pack}"` . By analyzing with C++ compilers directly, I think we can get the most accurate pack and size to compare.

Akarinnnnn committed:
SendSignal fixes without changing generated files (#9)

A workaround of incorrect codegen of `SteamNetConnectionInfo_t` .

Akarinnnnn committed:
Show pack and size of structs by C++

Akarinnnnn committed:
Pull from latest contributions

Akarinnnnn committed:
Merge branch 'sendsignal_gen_fixes' into anycpu-packable-fix-align-by-conditionalmarshal

Akarinnnnn committed:
Merge branch 'anycpu-packable-fix-align-by-conditionalmarshal' of https://github.com/Akarinnnnn/Steamworks.NET.AnyCPU into anycpu-packable-fix-align-by-conditionalmarshal

ropemonkey committed:
Fix invalid escape sequence warnings by converting regex patterns to raw strings

Akarinnnnn committed:
Merge branch 'anycpu-2' into anycpu-packable-fix-align-by-conditionalmarshal

Akarinnnnn committed:
Commit parser to be merged to packable branch

Akarinnnnn committed:
Merge branch 'sendsignal_gen_fixes' into anycpu-packable-fix-align-by-conditionalmarshal

Akarinnnnn committed:
Try to analyze structs like CGameID

Puropse is to exclude fake top level structs like CGameID::GameID_t. Maybe just add a table is enough?

Akarinnnnn commented:
Maybe it is too smart to my expectations for parser code. We just need to make it smart enough as git merge. I mean, some predefined special handling is enough just like old ways.

Akarinnnnn committed:
Add an comparer to official generated data.

Akarinnnnn committed:
Revert class parsing and fixed array related field handling

Akarinnnnn commented:
Bad news here, **I can't sync parser changes to my fork because git bugs**. I did things below:
0. Clone this repo if you haven't, it called Steamworks.NET below.
1. Prepare a bare repo storing my fork on your computer, steps below:
  a. `git init --bare SteamworksParser.git`
  b. `git remote add origin https://github.com/Akarinnnnn/SteamworksParser.git`
  c. `git fetch origin`
3. Attempt to sync changes to there by `git subtree push  '--prefix=CodeGen/SteamworksParser' ..\SteamworksParser.git smart-anycpu-marshal` on Steamworks.NET repo root,
4. Stuck on progress `17/71 (213) [199]`.

Would you guys meet this problem? This is very important for reviewing parser code changes. If you don't, would you help us to sync? I would gain write access of my fork to you.

Additionally, my filesystem storing code is ReFS, working on NTFS might not meet this problem.

Akarinnnnn commented:
@Chicken-Bones Incorrect marked pack aware structs should fixed a lot now. I've updated the [gist](https://gist.github.com/Akarinnnnn/0a35bb1d250505aa98020138498cfbe4) to including the list.

Akarinnnnn commented:
> Bad news here, **I can't sync parser changes to my fork because git bugs**. I did things below: 
> Additionally, my filesystem storing code is ReFS, working on NTFS might not meet this problem.

@Chicken-Bones Resolved, I pushed to GitHub directly on a NTFS drive, though some error messages thrown, it works.

Akarinnnnn commented:
Reopen this, it's accidently closed by GitHub bots.

Akarinnnnn committed:
Merge branch 'anycpu-2' into anycpu-packable-fix-align-by-conditionalmarshal

Akarinnnnn commented:
> Unfortunately this has turned out more complex than I envisioned. I'm willing to take a look at the parser myself but finding time will take longer. My apologies for the delay.

Do we accept every time Steamworks SDK updates we regenerate special struct table by C++ compiler? Or we just using this method to debug marshal bugs when it happens again?

Akarinnnnn committed:
Add a advanced script to test parser's pack-awareness

Akarinnnnn committed:
Regenerate after struct field scan algorithm update

Akarinnnnn committed:
Merge branch 'anycpu-packable-fix-align-by-conditionalmarshal' of https://github.com/Akarinnnnn/Steamworks.NET.AnyCPU into anycpu-packable-fix-align-by-conditionalmarshal

Akarinnnnn committed:
Changed DllImportResolver hack to allow searching .so etc. in assembly directory

Akarinnnnn committed:
Fixed old win64 steam native resolve logic

Akarinnnnn committed:
Fix large pack structs weren't being used

Akarinnnnn committed:
Merge upstream SteamworksParser changes

Akarinnnnn committed:
Minor misc changes

Update pre-baked generator resource.

Make `SteamMarshallerTable` public.

Added `.editorconfig` to enforce style.

sandsalamand commented:
It builds, but as soon as SteamManager.cs tries to load steam_api, it throws an error and crashes.

```
  ERROR: [Steamworks.NET] Could not load [lib]steam_api.dll/so/dylib. It's likely not in the correct location. Refer to the README for more details.
  ERROR: System.DllNotFoundException: Unable to load shared library 'steam_api' or one of its dependencies. In order to help diagnose loading problems, consider using a tool like strace. If you're using glibc, consider setting the LD_DEBUG environment variable: 
  ERROR: /usr/lib/dotnet/shared/Microsoft.NETCore.App/9.0.13/steam_api.so: cannot open shared object file: No such file or directory
  ERROR: /usr/lib/dotnet/shared/Microsoft.NETCore.App/9.0.13/libsteam_api.so: cannot open shared object file: No such file or directory
  ERROR: /usr/lib/dotnet/shared/Microsoft.NETCore.App/9.0.13/steam_api: cannot open shared object file: No such file or directory
  ERROR: /usr/lib/dotnet/shared/Microsoft.NETCore.App/9.0.13/libsteam_api: cannot open shared object file: No such file or directory
  ERROR: 
  ERROR:    at Steamworks.NativeMethods.SteamAPI_RestartAppIfNecessary(AppId_t unOwnAppID)
  ERROR:    at Steamworks.SteamAPI.RestartAppIfNecessary(AppId_t unOwnAppID)
  ERROR:    at SteamManager._EnterTree() in /home/sand/Documents/homunculus-fighters/MirageSteamworks/Runtime/SteamManager.cs:line 100<Node#27229423002>
```

Akarinnnnn commented:
Which branch areyou building aginast? I'm working on a super long named branch.



---Original---
From: ***@***.***&gt;
Date: Mon, Feb 23, 2026 15:27 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [Akarinnnnn/Steamworks.NET.AnyCPU] Fix pack size related marshalissues. (PR #1)


sandsalamand left a comment (Akarinnnnn/Steamworks.NET.AnyCPU#1)
 
It builds, but as soon as SteamManager.cs tries to load steam_api, it throws an error and crashes.
   ERROR: [Steamworks.NET] Could not load [lib]steam_api.dll/so/dylib. It's likely not in the correct location. Refer to the README for more details.   ERROR: System.DllNotFoundException: Unable to load shared library 'steam_api' or one of its dependencies. In order to help diagnose loading problems, consider using a tool like strace. If you're using glibc, consider setting the LD_DEBUG environment variable:    ERROR: /usr/lib/dotnet/shared/Microsoft.NETCore.App/9.0.13/steam_api.so: cannot open shared object file: No such file or directory   ERROR: /usr/lib/dotnet/shared/Microsoft.NETCore.App/9.0.13/libsteam_api.so: cannot open shared object file: No such file or directory   ERROR: /usr/lib/dotnet/shared/Microsoft.NETCore.App/9.0.13/steam_api: cannot open shared object file: No such file or directory   ERROR: /usr/lib/dotnet/shared/Microsoft.NETCore.App/9.0.13/libsteam_api: cannot open shared object file: No such file or directory   ERROR:    ERROR:    at Steamworks.NativeMethods.SteamAPI_RestartAppIfNecessary(AppId_t unOwnAppID)   ERROR:    at Steamworks.SteamAPI.RestartAppIfNecessary(AppId_t unOwnAppID)   ERROR:    at SteamManager._EnterTree() in /home/sand/Documents/homunculus-fighters/MirageSteamworks/Runtime/SteamManager.cs:line 100<Node#27229423002&gt;  
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***&gt;

sandsalamand commented:
I'm on `anycpu-packable-fix-align-by-conditionalmarshal`

Akarinnnnn commented:
I'll investgate it at 10:30 UTC+8.



---Original---
From: ***@***.***&gt;
Date: Tue, Feb 24, 2026 09:10 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [Akarinnnnn/Steamworks.NET.AnyCPU] Fix pack size related marshalissues. (PR #1)


sandsalamand left a comment (Akarinnnnn/Steamworks.NET.AnyCPU#1)
 
I'm on anycpu-packable-fix-align-by-conditionalmarshal
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***&gt;

sandsalamand commented:
Ok. In case this helps you, the steps that I followed to get that error above were:
1. Replace StandaloneV2.0 pre-built DLL from Steamworks main fork with the DLL that I built from your fork using `dotnet build Steamworks.NET.sln -c Release` within StandaloneV3.0.
2. Replace libsteam_api.so with the one from this fork.
3. Remove my NativeLibrary.SetDllImportResolver hack from SteamManager.cs.
4. Run the game.

sandsalamand commented on com.rlabrecque.steamworks.net/Runtime/autogen/NativeMethods.cs:
```diff
@@ -48,56 +48,58 @@ private static IntPtr DllImportResolver(string libraryName, System.Reflection.As
+                            extension = ".dylib";
+                        else
+                            extension = ".so"; // I can't imagine what else platforms other than linux that
+                                               // Steamworks.NET.AnyCPU will run on, but let's be future proof
```
You are adding the correct extension but not the `lib` prefix. On Linux and Mac the steam_api file is called `libsteam_api` .

sandsalamand replied:
I changed these lines to:

```
string prefix = "lib";	// The library is called libsteam_api on Linux and MacOS

// Will be libsteam_api.so on Linux and libsteam_api.dylib on MacOS
string unixLibraryName = prefix + SysPath.ChangeExtension(libraryName, extension);

string path = SysPath.Combine(assemblyLocation, unixLibraryName);
```

It did not solve the problem, but I confirmed that it is now the correct name "libsteam_api.so" on Linux.

Akarinnnnn commented:
> Ok. In case this helps you, the steps that I followed to get that error above were:
> 
> 1. Replace StandaloneV2.0 pre-built DLL from Steamworks main fork with the DLL that I built from your fork using `dotnet build Steamworks.NET.sln -c Release` within StandaloneV3.0.
> 2. Replace libsteam_api.so with the one from this fork.
> 3. Remove my NativeLibrary.SetDllImportResolver hack from SteamManager.cs.
> 4. Run the game.

It's little hard for me to follow your steps, especially I use WSL2 for Linux testing that possibly hard to run Godot editor. But I remember Godot uses standard .NET SDK to build scripts. It can help you copy steam .so files to proper location, but it also stops you from manual replace them.

Though your step 1 and 2 successfully replaced original V2 files, it might still being overwritten by .NET SDK. Try modify your `.csproj` like this might help:
[PackagedTest.zip](https://github.com/user-attachments/files/25507112/PackagedTest.zip)

Akarinnnnn commented:
Notably this line:
```xml
<PackageReference Include="Steamworks.NET.AnyCPU" Version="2025.162.7-a-socketmarshal.1" />
```

Replace `Steamworks.NET`(V2) with `Steamworks.NET.AnyCPU`(V3), and specify a correct version, your game will probably run @sandsalamand .

sandsalamand commented on com.rlabrecque.steamworks.net/Runtime/autogen/NativeMethods.cs:
```diff
@@ -48,56 +48,58 @@ private static IntPtr DllImportResolver(string libraryName, System.Reflection.As
+                            System.Diagnostics.Debug.WriteLine("It seems you are loading Steamworks.NET.AnyCPU from memory," +
+                                " auto-detect steam native location is not possible," +
+                                " please manually load steam natives into correct ALC.");
+                        }
```
This condition is always triggering for me on Linux.

Akarinnnnn replied:
Now I understand why, probably Godot didn't pass the file path to ALC loader API, therefore `Assembly.Location` is always not present. I will check source for this.

Akarinnnnn replied:
This is why. Godot needs special handling on `.so`.

https://github.com/godotengine/godot/blob/2327a823578a30f09068f97272598521896d5633/modules/mono/glue/GodotSharp/GodotPlugins/PluginLoadContext.cs#L27-L29

sandsalamand replied:
Wow, nice job investigating that. One question though, can't you just do what I did in SteamManager with `AppDomain.CurrentDomain.BaseDirectory` to get the Steamwork.NET.Dll executing folder?

Akarinnnnn replied:
Is it works? I'll align with your solution. Hope it works for you.

Akarinnnnn replied:
Try apply this patch to your source code:
```patch
diff --git a/Standalone3.0/autogen/NativeMethods.AnyCPU.cs b/Standalone3.0/autogen/NativeMethods.AnyCPU.cs
index 964c07a..1c59a9a 100644
--- a/Standalone3.0/autogen/NativeMethods.AnyCPU.cs
+++ b/Standalone3.0/autogen/NativeMethods.AnyCPU.cs
@@ -51,15 +51,19 @@ namespace Steamworks
                                                        extension = ".so"; // I can't imagine what else platforms other than linux that
                                                                                           // Steamworks.NET.AnyCPU will run on, but let's be future proof

-                                               string assemblyLocation = assembly.Location;
+                                               string searchDirectory = Path.GetDirectoryName(assembly.Location);

-                                               if (assemblyLocation == "") {
+                                               if (searchDirectory == "") {
                                                        System.Diagnostics.Debug.WriteLine("It seems you are loading Steamworks.NET.AnyCPU from memory," +
                                                                " auto-detect steam native location is not possible," +
-                                                               " please manually load steam natives into correct ALC.");
+                                                               " now trying to load from AppDomain.BaseDirectory." +
+                                                               " If still fails, please call" +
+                                                               " NativeLibrary.SetDllImporterResplver(typeof(Steamworks.SteamAPI).Assembly, YourResolver) manually.");
+
+                                                       searchDirectory = AppDomain.CurrentDomain.BaseDirectory;
                                                }

-                                               string path = Path.Combine(assemblyLocation, Path.ChangeExtension(nixPrefix + libraryName, extension));
+                                               string path = Path.Combine(searchDirectory, Path.ChangeExtension(nixPrefix + libraryName, extension));

                                                // second chance or user specified behavior search, not caring failures anymore
                                                NativeLibrary.TryLoad(path, assembly, searchPath, out lib);
```

sandsalamand replied:
Okay, I tested that patch. You need to change `if (searchDirectory == "")` to `if (string.IsNullOrEmpty(searchDirectory))`, but aside from that, it works.

Akarinnnnn replied:
Thanks for testing, I'll change the condition for correctness.

sandsalamand reviewed (commented):
Possible causes of the error

ryan-linehan commented:
> Notably this line:
> 
> ```xml
> 
> <PackageReference Include="Steamworks.NET.AnyCPU" Version="2025.162.7-a-socketmarshal.1" />
> 
> ```
> 
> 
> 
> Replace `Steamworks.NET`(V2) with `Steamworks.NET.AnyCPU`(V3), and specify a correct version, your game will probably run @sandsalamand .

I've got a Godot game building and ran it on a steam deck through steam with version 2025.162.6.b-socket.1 if it helps. I can try updating to this one to make sure it works on my demo project if it helps out at all

Akarinnnnn committed:
Add missing "lib" prefix to native resolve logic

Move Any CPU native resolve logic to another file.

Minor changes to .editorconfig

Akarinnnnn commented:
> > Notably this line:
> > ```
> > <PackageReference Include="Steamworks.NET.AnyCPU" Version="2025.162.7-a-socketmarshal.1" />
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Replace `Steamworks.NET`(V2) with `Steamworks.NET.AnyCPU`(V3), and specify a correct version, your game will probably run @sandsalamand .
> 
> I've got a Godot game building and ran it on a steam deck through steam with version 2025.162.6.b-socket.1 if it helps. I can try updating to this one to make sure it works on my demo project if it helps out at all

Your game is working right? Which version of Godot are you two using? There might be some version caused differences.

Akarinnnnn committed:
Fix Godot loading compatibility
